torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,366 @@
+                self._dist_broadcast_coalesced(self.modules_buffers_data[0],
+                                               self.broadcast_bucket_size)
+                if len(self.device_ids) > 1:
+                    # intra-node buffer sync
+                    result = broadcast_coalesced(self.modules_buffers_data[0],
+                                                 self.device_ids,
+                                                 self.broadcast_bucket_size)
+                    for tensors, module_buffers_data in zip(result[1:], self.modules_buffers_data[1:]):
+                        for tensor, buffer_data in zip(tensors, module_buffers_data):
+                            buffer_data.set_(tensor)
+
+    def _register_grad_hooks(self):
+        self._grad_accs = []  # need to keep them in scope
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                if p.requires_grad:
+                    p_tmp = p.expand_as(p)
+                    grad_acc = p_tmp.grad_fn.next_functions[0][0]
+                    grad_acc.register_hook(self._make_param_hook(p, device_idx))
+                    self._grad_accs.append(grad_acc)
+
+    def _make_param_hook(self, param, device_idx):
+        bucket_idx = self.bucket_map[param]
+
+        def distributed_data_parallel_hook(*unused):
+            if param.grad.requires_grad:
+                raise RuntimeError("DistributedDataParallelC10d only works "
+                                   "with gradients that don't require grad")
+            bucket = self.buckets[bucket_idx][device_idx]
+            bucket.append(param.grad.data)
+
+            # We can flush these and save memory for replicas
+            if device_idx > 0:
+                param.grad = None
+                param.data.set_()
+
+            # Current device's bucket is full
+            if len(bucket) == self.bucket_sizes[bucket_idx]:
+                self.devs_ready[bucket_idx] += 1
+                if self.devs_ready[bucket_idx] < len(self.device_ids):
+                    return
+
+                if bucket_idx not in self.buckets_ready:
+                    self.buckets_ready.add(bucket_idx)
+                    if len(self.buckets_to_reduce) > 0 and self.buckets_to_reduce[-1] == bucket_idx:
+                        # Reduce the bucket
+                        self._queue_reduction(bucket_idx)
+                        self.buckets_to_reduce.pop()
+
+                    # If all buckets are ready
+                    if len(self.buckets_ready) == len(self.bucket_order):
+                        # In case there are unreduced buckets, reduce them all
+                        if len(self.buckets_to_reduce) > 0:
+                            for b_idx in reversed(self.buckets_to_reduce):
+                                self._queue_reduction(b_idx)
+                        # A final sync for all the reduction works
+                        self._sync_reduction_works()
```

Comment:
Ok, GitHub has closed the previous thread, but I can't see how my proposal violates the ordering (happy to see a proof that it doesn't work, but I'm pretty sure it does). IIRC this is exactly what DDP does today. There's no reason to halt *any* processing of buckets until the very end of backward just because *one* of them came early. You should only make it wait until all previous ones are ready, and then trigger all reductions that you can once this happens. This is what the `for` loop would let you achieve, with little modification to the current code.

Reply:
addressed